### DISCUSSION TITLE
Setting icon for new keymaps

### DISCUSSION THREAD

**Discussion by Irubataru** [LazyVim/LazyVim#4512] [Q&A] [answered]
I am trying to add some extra keymaps to copilot, but I can't seem to get it working, and don't know what is wrong. I have the following file in my plugins folder:

```lua
return {
  {
    "zbirenbaum/copilot.lua",
    keys = {
      {
        "<leader>at",
        function()
          require("copilot.command").toggle()
        end,
        desc = "Toggle copilot (buffer)",
        icon = LazyVim.config.icons.kinds.Copilot,
        mode = { "n", "v" },
      },
      {
        "<leader>aT",
        function()
          local is_disabled = require("copilot.client").is_disabled()

          if is_disabled then
            require("copilot.command").enable()
          else
            require("copilot.command").disable()
          end
        end,
        desc = "Toggle copilot (global)",
        icon = { icon = LazyVim.config.icons.kinds.Copilot, color = "orange" },
        mode = { "n", "v" },
      },
    },
  },
}
```

The keymaps work, but they have the wrong icons

![image](https://github.com/user-attachments/assets/50e19da7-7d36-46d8-a59f-f5fe43a3a16c)

**Comment by dpetka2001** [comment]
This is not configurable at the moment. You would have to use `LazyVim.toggle.map` with some slight changes in the code to make this happen, which is currently not possible.

**Reply by Irubataru** [reply]
Aww, that's a bummer. So I should just configure them in `opts` then, adding keymaps with which-key directly instead of using the `keys` table?

Does anyone know where it is specified that the CopilotChat keybinds get the copilot icon? I can't find it anywhere in the LazyVim source.

**Reply by jyuan0** [reply]
> Does anyone know where it is specified that the CopilotChat keybinds get the copilot icon? I can't find it anywhere in the LazyVim source.

I think it's here: https://github.com/folke/which-key.nvim/blob/main/lua/which-key/icons.lua#L31

**Comment by jyuan0** [comment, marked as answer]
To change the icons, a which-key icon rule might work:
```lua
  {
    "folke/which-key.nvim",
    opts = function(_, opts)
      opts.icons = {
        rules = {
          { pattern = "copilot", icon = LazyVim.config.icons.kinds.Copilot, color = "orange" }
        }
      }
    end,
  }
```

**Reply by dpetka2001** [reply]
This won't be dynamic like the current implementation of `LazyVim.toggle.map` is though.

**Reply by dpetka2001** [reply]
I've made some really minor changes to `LazyVim.toggle.wk` to take into consideration user's provided icons and colors. But I really don't know if something like this would be desirable by Folke. 

Should I make a PR about it? Is this something that users would want @jyuan0 in your opinion? 

But that would also mean that someone would have to get accustomed to using `LazyVim.toggle.map`. I don't believe that many people would use that? Just guessing here.

**Reply by jyuan0** [reply]
> Should I make a PR about it? Is this something that users would want @jyuan0 in your opinion?

I don't know how many users would be interested, but if you have the time and desire, I think it would be nice. The last time I looked, it was a bit cumbersome to make any changes to the which-key icons and colors.

> But that would also mean that someone would have to get accustomed to using LazyVim.toggle.map. I don't believe that many people would use that? Just guessing here.

Not sure either, but if I understand you correctly, asking users to switch to LazyVim.toggle.map if they want dynamic icons and colors doesn't seem unreasonable to me.

**Reply by dpetka2001** [reply]
Well, I had it ready already. The additions were really minor to the current `LazyVim.toggle.wk` function. 

@jyuan0 If you could test out the PR #4521 locally, I'd appreciate it. 

As an example I added in `/lua/config/keymaps.lua` the following
```lua
LazyVim.toggle.map("<leader>uz", {
  name = "MyRelativeNumber",
  icon = LazyVim.config.icons.kinds.Copilot,
  color_enabled = "blue",
  color_disabled = "orange",
  get = function()
    return vim.opt_local.relativenumber:get() == true
  end,
  set = function(state)
    vim.opt_local.relativenumber = state and true or false
  end,
})
```
Please bear in mind this can't be used in `keys` (LazyKeysHandler), at least that's what I believe although I might be wrong (hopefully Folke or someone else can correct me if I'm wrong about this). Users would either have to add it in `/lua/config/keymaps.lua` or in the `opts` or `config` function of the plugin spec they're interested in like the rest of usages of `LazyVim.toggle.map` throughout the codebase currently.

**Reply by jyuan0** [reply]
> @jyuan0 If you could test out the PR https://github.com/LazyVim/LazyVim/pull/4521 locally, I'd appreciate it.

The code works locally for me. I changed one of my toggles in `keymaps.lua` and the icon is updated, and the colors change depending on state.

**Reply by dpetka2001** [reply]
Cool!! Tnx for testing it. Let's hope it's something that Folke is willing to merge :smile:

**Reply by Irubataru** [reply]
I guess I can use the which-key `add` function to add the icon instead of simply using the `keys` table 🤔 Thanks for the reply!

**Reply by jyuan0** [reply]
Yes, using which-key `add` would achieve the same. You'd lose the plugin lazy-loading, but that doesn't matter here. It would be nice if you could set icons in `keys`, but I don't think that's possible yet.

**Reply by Irubataru** [reply]
I went with the which-key pattern in the end, it worked the best. Thanks for the suggestion 😊

Hope they add the ability to specify it in the icon in the keys table eventually!

**Reply by dpetka2001** [reply]
I wouldn't hold high hopes for this to happen (although I might get proven wrong in the future and I would welcome such a change to be honest), since I'm guessing that this is left up to plugins to implement that (which-key in this case). The best way to do what you want right now is to either use which-key `rules` if you want static icons or use `LazyVim.toggle.map` if you want dynamic icons and colors based on state.
